CentOS 7 下编译出错

环境：CentOS 7.1 x64， 系统自带 Python 没有安装任何额外包。已编译最新版 OpenBLAS。无 ICC & MKL。使用
```shell
cmake  .. -DWITH_GPU=OFF 
```
进行配置和编译。

前面出现过若干小问题，但都还可以解决。直到此报错：
```cmake
[ 33%] Building CXX object paddle/math/CMakeFiles/paddle_math.dir/BaseMatrix.cu.o
In file included from /home/enigma/Paddle-develop/paddle/math/BaseMatrix.cu:23:0:
/home/enigma/Paddle-develop/paddle/math/MathFunctions.h:30:21: fatal error: lapacke.h: No such file or directory
```

在 [Build from Source](http://www.paddlepaddle.org/doc/build/build_from_source.html) 中并没有提到 Paddle 需要 LAPACK 。此问题应该如何解决？

以及 『前面出现过若干的小问题』 里包括一个，在 build 目录下 `cmake ..` 的时候会出现：
```shell
CMake Error at cmake/external/protobuf.cmake:32 (IF):
  if given arguments:

    "STREQUAL" "centos"

  Unknown arguments specified
Call Stack (most recent call first):
  CMakeLists.txt:62 (include)


-- Configuring incomplete, errors occurred!
```
这样的报错。我是注释掉 `cmake/external/protobuf.cmake` 中的第  32 34 35 36 行来临时解决这个问题的。这可能是由于什么引起的？

@gangliao 为 third_party 中的 python 安装一些包的时候会发生类似的问题：
```
[  6%] Performing build step for 'numpy'
CMake Error at /home/enigma/src_packages/Paddle-develop/third_party/python/numpy/src/numpy-stamp/numpy-build-RelWithDebInfo.cmake:16 (message):
  Command failed: 1

   'env' 'PATH=/home/enigma/src_packages/Paddle-develop/third_party/install/python/bin' 'PYTHONHOME=/home/enigma/src_packages/Paddle-develop/third_party/install/python' 'PYTHONPATH=/home/enigma/src_packages/Paddle-develop/third_party/install/python/lib:/home/enigma/src_packages/Paddle-develop/third_party/install/python/lib/python2.7:/home/enigma/src_packages/Paddle-develop/third_party/install/python/lib/python2.7/site-packages' '/home/enigma/src_packages/Paddle-develop/third_party/install/python/bin/python' 'setup.py' 'build'

  See also

    /home/enigma/src_packages/Paddle-develop/third_party/python/numpy/src/numpy-stamp/numpy-build-*.log
```
可能与什么相关，以及应该如何解决？

@gangliao 我重新查了一次，没有装任何 Python 包的时候，CentOS 7 core 下的 cmake .. 是这样的输出：
```cmake
-- Found PythonInterp: /usr/bin/python2.7 (found suitable version "2.7.5", minimum required is "2.7") 
-- Could NOT find PythonLibs (missing:  PYTHON_LIBRARIES PYTHON_INCLUDE_DIRS) (Required is at least version "2.7")
[Paddle] Python Executable: /home/enigma/src_packages/Paddle-develop/third_party/install/python/bin/python
[Paddle] Python Include: 
[Paddle] Python Libraries: /home/enigma/src_packages/Paddle-develop/third_party/install/python/lib/libpython2.7.a
```
也就是找到了 PythonInterp 但是没有找到 PythonLibs。用 `yum install python-devel` 以后再 cmake .. ，输出变成了：
```cmake
-- Found PythonInterp: /usr/bin/python2.7 (found suitable version "2.7.5", minimum required is "2.7") 
-- Found PythonLibs: /usr/lib64/libpython2.7.so (found suitable version "2.7.5", minimum required is "2.7") 
-- Could NOT find PY_pip (missing:  PY_PIP) 
CMake Error at cmake/python_module.cmake:27 (message):
  python module pip is not found
Call Stack (most recent call first):
  cmake/external/python.cmake:24 (find_python_module)
  CMakeLists.txt:63 (include)
```
安装 pip 并用 pip install numpy protobuf 以后可以检查通过。

@gangliao 
有两个centos7出现的情况可能需要反应一下
第一个是cmake/external/protobuf.cmake里的ExternalProject_Add中可能需要加上zlib的参数，否则在环境里没有zlib的机器上，自动下载编译的protobuf会找到不到zlib，从而缺少zlib的支持
```
57     -Dprotobuf_BUILD_TESTS=OFF
58     -DZLIB_ROOT:FILEPATH=${ZLIB_ROOT} #或许需要加上这一行
59     -DCMAKE_POSITION_INDEPENDENT_CODE=ON
```

第二个是cmake时，cmake/external/protobuf.cmake里会出现这样的错误
```
CMake Error at cmake/external/protobuf.cmake:32 (IF):
  if given arguments:

    "STREQUAL" "centos"

  Unknown arguments specified
```
在把32行的${HOST_SYSTEM}加上引号以后问题解决，不知道是不是cmake的版本问题
```
32   IF("${HOST_SYSTEM}" STREQUAL "centos")
```

--- a/cmake/system.cmake
+++ b/cmake/system.cmake
@@ -21,6 +21,7 @@ ELSE(WIN32)
         SET(MACOS_VERSION ${VERSION})
         SET(HOST_SYSTEM "macosx")
     ELSE(APPLE)
+
         IF(EXISTS "/etc/issue")
             FILE(READ "/etc/issue" LINUX_ISSUE)
             IF(LINUX_ISSUE MATCHES "CentOS")
@@ -31,6 +32,14 @@ ELSE(WIN32)
                 SET(HOST_SYSTEM "ubuntu")
             ENDIF()
         ENDIF(EXISTS "/etc/issue")
+
+        IF(EXISTS "/etc/redhat-release")
+            FILE(READ "/etc/redhat-release" LINUX_ISSUE)
+            IF(LINUX_ISSUE MATCHES "CentOS")
+                SET(HOST_SYSTEM "centos")
+            ENDIF()
+        ENDIF(EXISTS "/etc/redhat-release")
+
     ENDIF(APPLE)
 ENDIF(WIN32)
 

